### PR TITLE
Fix faulty Spiro handling in SplinePointListSet

### LINKED ISSUE #4037: Crash while scaling spline (perhaps Spiro related)
2019-11-28-193852_1978x1121_scrot

1. Open [53.sfd](https://github.com/fontforge/fontforge/files/3901432/53.zip)
2. Open glyph `five`
3. Verify Spiro mode is enabled
4. Select all points in the digit five by clicking any of its points
4. Scale tool
5. Start scale at corner point highlighted in image
6. Scale size to match size of smaller three
7. You should get the following, in my experience highly reliable, crash:

```gdb
Program received signal SIGSEGV, Segmentation fault.
0x0000155555104f8c in SplinePointListSet (tobase=0x555555dc8840, 
    frombase=0x555555f4f760) at ../fontforge/splineutil2.c:3648
3648		for ( tsp = tobase->first, fsp=frombase->first; tsp!=first; tsp=tsp->next->to, fsp=fsp->next->to ) {
(gdb) bt
#0  0x0000155555104f8c in SplinePointListSet (tobase=0x555555dc8840, frombase=0x555555f4f760) at ../fontforge/splineutil2.c:3648
#1  0x0000155554e454e5 in _CVRestoreTOriginalState (cv=0x15554ed48010, p=0x15554ed4b3d0) at ../fontforge/cvundoes.c:1080
#2  0x000055555567e51f in CVRestoreTOriginalState (cv=0x15554ed48010) at ../fontforgeexe/cvpointer.c:549
#3  0x00005555556a5957 in CVMouseMoveTransform (cv=0x15554ed48010) at ../fontforgeexe/cvtranstools.c:44
#4  0x0000555555603c5f in CVMouseMove (cv=0x15554ed48010, event=0x7fffffffd130) at ../fontforgeexe/charview.c:5422
#5  0x0000555555604fe8 in v_e_h (gw=0x555555f048f0, event=0x7fffffffd130) at ../fontforgeexe/charview.c:5732
#6  0x000055555584e5bc in _GWidget_Container_eh (gw=0x555555f048f0, event=0x7fffffffd130) at ../gdraw/gcontainer.c:405
#7  0x0000555555862885 in _GGDKDraw_CallEHChecked (gw=0x555555f048f0, event=0x7fffffffd130, eh=0x55555584d91c <_GWidget_Container_eh>) at ../gdraw/ggdkdraw.c:334
#8  0x00005555558652a8 in _GGDKDraw_DispatchEvent (event=0x555555f1a320, data=0x555555be1a40) at ../gdraw/ggdkdraw.c:1237
#9  0x0000155554195954 in  () at /usr/lib/libgdk-3.so.0
#10 0x0000155554142904 in  () at /usr/lib/libgdk-3.so.0
#11 0x0000155553d1639e in g_main_context_dispatch () at /usr/lib/libglib-2.0.so.0
#12 0x0000155553d181b1 in  () at /usr/lib/libglib-2.0.so.0
#13 0x0000155553d181f1 in g_main_context_iteration () at /usr/lib/libglib-2.0.so.0
#14 0x0000555555868052 in GGDKDrawEventLoop (gdisp=0x555555be1a40) at ../gdraw/ggdkdraw.c:2204
#15 0x0000555555853ada in GDrawEventLoop (gdisp=0x555555be1a40) at ../gdraw/gdraw.c:789
#16 0x0000555555817e7b in fontforge_main (argc=2, argv=0x7fffffffe568) at ../fontforgeexe/startui.c:1415
#17 0x00005555555aab89 in main (argc=2, argv=0x7fffffffe568) at ../fontforgeexe/main.c:33
```

### PR DESCRIPTION
**Pull request by ctrlcctrlv**
This closes #4037.

### CODE REVIEW

File: fontforge/splineutil2.c [hunk 1]
```diff
@@ -3654,12 +3654,12 @@ void SplinePointListSet(SplinePointList *tobase, SplinePointList *frombase) {
 	    tsp->noprevcp = fsp->noprevcp;
 	    tsp->nonextcp = fsp->nonextcp;
 	    if ( first==NULL ) first = tsp;
-	    if ( tsp->next==NULL )
+	    if ( tsp->next==NULL || fsp->next==NULL )
```
Comment (frank-trampe) [inline fontforge/splineutil2.c]: What would make these deviate so as to require the double-check? (I know I can use the test file to get there, but I figure you have a more general understanding.)

File: fontforge/splineutil.c [hunk 2]
```diff
@@ -1945,6 +1945,7 @@ SplinePointList *SplinePointListTransformExtended(SplinePointList *base, real tr
 	    int i;
 	    for ( i=0; i<spl->spiro_cnt-1; ++i )
 		TransformSpiro(&spl->spiros[i], transform);
+		SSRegenerateFromSpiros(spl);
```
Comment (frank-trampe) [inline fontforge/splineutil.c]: Was this omission breaking all transformations of spiros?

Comment (ctrlcctrlv) [reply fontforge/splineutil.c]: It was making them look really weird in the UI; not complete breakage, but close.

File: fontforge/splineutil2.c [hunk 3]
```diff
@@ -3654,12 +3654,12 @@ void SplinePointListSet(SplinePointList *tobase, SplinePointList *frombase) {
...
 	break;
 	}
 	firstsp = NULL;
 	for ( tspline = tobase->first->next, fspline=frombase->first->next;
-		tspline!=NULL && tspline!=firstsp;
+		tspline!=NULL && tspline!=firstsp && frombase->spiro_cnt==0;
```
Comment (frank-trampe) [inline fontforge/splineutil2.c]: I don't understand the point of the new condition. Does it change during the iterations?

File: fontforge/splineutil.c (hunk 2 above)
Comment (ctrlcctrlv) [reply fontforge/splineutil.c]: (We can see it was simply forgotten as it's used elsewhere, e.g. in function SplinePointListSpiroTransform, splineutil.c:2023)

**Comment by ctrlcctrlv** [comment]
Sorry, I have to close this. It solves an immediate crash but causes more problems elseswhere...I'll reopen it if I can figure out a better fix.

**Pull request closed without merging**